contracts/CrossChain.sol

```diff
@@ -322,9 +398,141 @@ contract CrossChain is System, ICrossChain, IParamSubscriber{
         registeredContractChannelMap[handlerContract][channelId] = isEnable;
         emit enableOrDisableChannel(channelId, isEnable);
       }
+    } else if (Memory.compareStrings(key, "emergencySuspendApproveThreshold")) {
+      require(value.length == 2, "length of value for emergencySuspendApproveThreshold should be 2");
+      uint16 emergencySuspendApproveThreshold = BytesToTypes.bytesToUint16(32, value);
+      require(emergencySuspendApproveThreshold > 0, "zero suspend threshold");
+      approveThresholdMap[EMERGENCY_SUSPEND_PROPOSAL] = emergencySuspendApproveThreshold;
+    } else if (Memory.compareStrings(key, "reopenApproveThreshold")) {
+      require(value.length == 2, "length of value for reopenApproveThreshold should be 2");
+      uint16 reopenApproveThreshold = BytesToTypes.bytesToUint16(32, value);
+      require(reopenApproveThreshold > 0, "zero reopen threshold");
+      approveThresholdMap[REOPEN_PROPOSAL] = reopenApproveThreshold;
+    } else if (Memory.compareStrings(key, "cancelTransferApproveThreshold")) {
+      require(value.length == 2, "length of value for cancelTransferApproveThreshold should be 2");
+      uint16 cancelTransferApproveThreshold = BytesToTypes.bytesToUint16(32, value);
+      require(cancelTransferApproveThreshold > 0, "zero cancel transfer threshold");
+      approveThresholdMap[CANCEL_TRANSFER_PROPOSAL] = cancelTransferApproveThreshold;
     } else {
       require(false, "unknown param");
     }
     emit paramChange(key, value);
   }
-}
\ No newline at end of file
+
+  // BEP-171: Security Enhancement for Cross-Chain Module
+  function challenge(
+    // to avoid stack too deep error, using `uint64[4] calldata params`
+    // instead of  `uint64 height0, uint64 height1, uint64 packageSequence, uint8 channelId`
+    uint64[4] calldata params, // 0-height0, 1-height1, 2-packageSequence, 3-channelId,
+    bytes calldata payload0,
+    bytes calldata payload1,
+    bytes calldata proof0,
+    bytes calldata proof1
+  )
+  onlyInit
+  blockSynced(params[0])
+  blockSynced(params[1])
+  channelSupported(uint8(params[3]))
+  whenNotSuspended
+  external {
+    // the same key with different values (payloads)
+    require(keccak256(payload0) != keccak256(payload1), "same payload");
+
+    bytes memory _key;
+    uint64 _packageSequence;
+    uint8 _channelId;
+    {
+      _packageSequence = params[2];
+      _channelId = uint8(params[3]);
+      _key = generateKey(_packageSequence, _channelId);
+      bytes32 _keyHash = keccak256(_key);
+      require(!challenged[_keyHash], "already challenged");
+
+      // if succeeding in challenge
+      challenged[_keyHash] = true;
```

Comment:
Can we move this logic down?

Reply:
will result in `Stack too deep, try removing local variables.`